### ISSUE TITLE
wasm-bindgen-futures Queue fails in Firefox

### ISSUE THREAD

**Issue by MartinKavik** [rustwasm/wasm-bindgen#3635]
### Describe the Bug

A browser app using `wasm-bindgen` panics in **Firefox** when `window.open` is run together with `spawn_local`.

### Steps to Reproduce

Minimal example:
```rust
#[wasm_bindgen(start)]
pub async fn main_js() {
    spawn_local(async {});   
    let _ = web_sys::window().unwrap_throw().open_with_url("http://example.com");
}
```

I've prepared a demo based on the Dominator framework:
1. See the code https://github.com/MartinKavik/rust-dominator/blob/firefox_queue_issue_demo/examples/issue/src/lib.rs
2. Clone the repo
3. `cd examples/issue`
4. `yarn install`
5. `yarn watch`
6. Open http://localhost:10001 in Firefox (tested on Windows and Linux)
7. Allow pop-ups if needed (and then refresh the tab if needed as well)
8. Open DevTools
9. Notice the error `closure invoked recursively or after being dropped` in the Console

### Expected Behavior

No panic.

### Actual Behavior

`wasm-bindgen-futures`'s `Queue.closure` is accidentally called recursively and then it panics because it's not permitted to invoke `Closure<dyn FnMut(JsValue)>` recursively.
Edit `Cargo.toml` in the demo - uncomment the dependencies marked by the comment `# Firefox queue issue debug` and run `yarn watch` again.  Refresh your Firefox tab and notice the logs ordering in the Console:

```
Closure started! 
Open the window!
Closure started!
From task! 
Closure stopped!
Main End!
Closure stopped!
```
and compare it with the output in Chrome:

```
Closure started!
Open the window!
Main End!
Closure stopped!
Closure started!
From task!
Closure stopped!
```
(The only change in the code is replaced `Queue.closure` with new `Closure`s to work around the `FnMut` nested calls limitation.)

My wild guess is that `window.open` call somehow breaks Firefox microtask handling (maybe because it's probably partly async, waiting for browser/user input?) and then both `promise.then(..)` and `queueMicrotask(..)` don't work as expected.

### Additional Context

The code doesn't panic in the older `wasm-bindgen-futures` (**<= 0.4.29**) because its `Queue` implementation is greedy so it's not possible to accidentally schedule a new task execution pipeline while the previous one is still running. 

Also I've written a possible solution.
You can try it out by changing the `Cargo.toml` dependencies to the ones marked with `# Firefox queue issue fix` or `# Firefox queue issue fix without logs`.
You can see the [code diff](https://github.com/rustwasm/wasm-bindgen/compare/main...MartinKavik:wasm-bindgen:firefox_queue_issue_fix) or only the [updated queue.rs](https://github.com/MartinKavik/wasm-bindgen/blob/firefox_queue_issue_fix/crates/futures/src/queue.rs).
The idea is to not use `Queue.promise` just as a dummy thing to call `then` on it as a poor man's `queueMicrotask` but to use it basically as an async alternative to the old `is_spinning` flag from the older `Queue` implementation to prevent nested `Closure` invocations from nested tasks. It means the `promise` isn't immediately resolved but it's resolved at the end of the `run_all` (tasks) method. That way we can no longer use `queueMicrotask` and we have to create a new `Promise` on every `run_all` call but I hope it will be good enough.
I can add some comments / polish the code as needed and create a PR if you want, just let me know.
Thanks!

**Comment by flumm** [comment]
sorry for the noise, but this piqued my interest but this seems to be generally broken in firefox (even with raw javascript) ? this code snippet:

```javascript
let inner = function() { 
    console.trace("inner");
};
let outer = function() {
    console.trace("outer");
    queueMicrotask(inner);
    window.open("http://google.com").close();
}; 
let outer2 = function() { 
    console.trace("outer2");
    queueMicrotask(inner);
};
queueMicrotask(outer);  
queueMicrotask(outer2);
````

pasted into the dev console produces following output:

> console.trace() outer debugger eval code:1:86
>     outer debugger eval code:1
>     (Async: VoidFunction)
>     \<anonymous\> debugger eval code:1
>
> console.trace() outer2 debugger eval code:1:203
>     outer2 debugger eval code:1
>     outer debugger eval code:1
>     (Async: VoidFunction)
>     \<anonymous\> debugger eval code:1
>
> console.trace() inner debugger eval code:1:34
>     inner debugger eval code:1
>     outer debugger eval code:1
>     (Async: VoidFunction)
>     \<anonymous\> debugger eval code:1
>
> console.trace() inner debugger eval code:1:34
>     inner debugger eval code:1
>     outer debugger eval code:1
>     (Async: VoidFunction)
>     \<anonymous\> debugger eval code:1

so it seems that outer2 is called in context of outer, and both inner calls as well ? (which imho does not make sense)

in chrome the output is:

> 0:07:27.770 VM70:1 outer
> outer @ VM70:1
> 10:07:27.770 VM70:1 outer2
> outer2 @ VM70:1
> 10:07:27.770 VM70:1 inner
> inner @ VM70:1
> 10:07:27.770 VM70:1 inner
> inner @ VM70:1

so each task has it's own call stack without any nesting? (which intuitively seems correct)

maybe this helps

**Comment by daxpedda** [comment]
> so it seems that outer2 is called in context of outer, and both inner calls as well ? (which imho does not make sense)

This doesn't happen to me in Firefox 118, which is fairly recent so they might have fixed something:
```
console.trace() outer
    outer debugger eval code:5
    (Async: VoidFunction)
    <anonymous> debugger eval code:13

console.trace() outer2
    outer2 debugger eval code:10
    (Async: VoidFunction)
    <anonymous> debugger eval code:14

console.trace() inner
    inner debugger eval code:2
    (Async: VoidFunction)
    outer debugger eval code:6
    (Async: VoidFunction)
    <anonymous> debugger eval code:13

console.trace() inner
    inner debugger eval code:2
    (Async: VoidFunction)
    outer2 debugger eval code:11
    (Async: VoidFunction)
    <anonymous> debugger eval code:14
```
The code MartinKavik posted is still broken though, so there is still an issue.
In any case, both my output and @flumm doesn't actually indicate that a task is being called twice, which is the problem in the Rust code.

**Comment by flumm** [comment]
> This doesn't happen to me in Firefox 118, which is fairly recent so they might have fixed something:

well i tested this too on 118 (Linux though) and that was my behaviour...

> The code MartinKavik posted is still broken though, so there is still an issue.
> In any case, both my output and @flumm doesn't actually indicate that a task is being called twice, which is the problem in the Rust code.

AFAICS the issue is that the closure which starts 'run_all' is called recursively in that case because of the following events (please correct me if i'm wrong):

1. main is called async with 'spawn_local' wich queues the 'run_all' closure
2. the queued task is run which again calls 'spawn_local' to queue the inner task (which again queues the 'run_all' closure)
3. while the task is still running, window.open().close() is running, messing up the firefox task handling
4. then for some reason, the second queued run_all closure is started, **but in the context of the still running first closure** -> error is triggered

which would be explained by the stack traces of my example code (honestly no idea why it shows different stack traces than my browser)

**Comment by daxpedda** [comment]
> well i tested this too on 118 (Linux though)

I'm on Linux as well.

> 4\. then for some reason, the second queued run_all closure is started, **but in the context of the still running first closure** -> error is triggered (running is probably the wrong word, but it still exists as the inner function is called in it's context)

I mean it has to actually **run** before the closure finishes to trigger the error.

---

I actually just tried a really minimal example:
```js
let inner = () => console.log("inner")
let outer = () => {
    console.log("outer");
    queueMicrotask(inner);
    window.open("http://google.com").close();
    console.log("outer finish")
}
queueMicrotask(outer)
```
Which results in:
```
outer
inner
outer finish
```
This is it, somehow `inner` is called before `outer` finishes. I will write up a bug report in Firefox.

**Comment by daxpedda** [comment]
I also found the bug that explains why this was happening before we started using `queueMicrotask()`: https://bugzilla.mozilla.org/show_bug.cgi?id=1294029.
This is the new bug report I filed: https://bugzilla.mozilla.org/show_bug.cgi?id=1855409.

@MartinKavik the workaround you posted is a bit too heavy-handed for me and I'm not sure I want to reduce the already abysmal performance of Rust async in Firefox even further, not mentioning that this also affects Chrome, only to circumvent a bug around `Window.open()`, which I assume is a tiny fraction of `spawn_local()` usage.

Therefor I will close this for now and hopefully Firefox will fix this.
